File: model/meal.go
package model

import (
	"time"

	"gorm.io/gorm"
)

type MealType int

const (
	MealTypeBreakfast MealType = 1
	MealTypeLunch     MealType = 2
	MealTypeDinner    MealType = 3
	// MealTypeSnack     MealType = 4
)

func (m MealType) String() string {
	switch m {
	case MealTypeBreakfast:
		return "早餐"
	case MealTypeLunch:
		return "午餐"
	case MealTypeDinner:
		return "晚餐"
	// case MealTypeSnack:
	// 	return "零食"
	default:
		return "未知"
	}
}

type Meal struct {

	// ID        uint       `gorm:"primaryKey;autoIncrement" json:"id"`
	gorm.Model
	UserID    uint      `json:"user_id"`
	FoodName  string    `gorm:"type:varchar(100);not null" json:"food_name"`
	MealType  string    `gorm:"type:varchar(10);not null" json:"meal_type"`
	MealDate  time.Time `gorm:"type:date;not null" json:"meal_date"`
	ImagePath string    `gorm:"type:varchar(255)" json:"image_path,omitempty"`
	Calories  float64   `gorm:"type:decimal(8,1);default:0" json:"calories"`
	ProteinG  float64   `gorm:"type:decimal(6,1);default:0" json:"protein_g"`
	FatG      float64   `gorm:"type:decimal(6,1);default:0" json:"fat_g"`
	CarbsG    float64   `gorm:"type:decimal(6,1);default:0" json:"carbs_g"`
	// CreatedAt time.Time  `gorm:"autoCreateTime" json:"created_at,omitempty"`
	// UpdatedAt time.Time  `gorm:"autoUpdateTime" json:"updated_at,omitempty"`
	// DeletedAt *time.Time `gorm:"index" json:"deleted_at,omitempty"`
}

func (*Meal) TableName() string {
	return "meals"
}

type FoodStatistic struct {
	FoodName string `json:"food_name"`
	Count    int64  `json:"count"`
}

type NutritionSummary struct {
	Calories float64 `json:"calories"`
	ProteinG float64 `json:"protein_g"`
	FatG     float64 `json:"fat_g"`
	CarbsG   float64 `json:"carbs_g"`
}

File: frontend/src/services/meal.ts
import api from './api'

export interface Meal {
  ID: number
  user_id: number
  food_name: string
  meal_type: string
  meal_date: string
  image_path?: string
  calories: number
  protein_g: number
  fat_g: number
  carbs_g: number
  created_at?: string
  updated_at?: string
}

export interface CreateMealRequest {
  food_name: string
  meal_type: number
  meal_date: string
  image?: File
  calories?: number
  protein_g?: number
  fat_g?: number
  carbs_g?: number
}

export interface UpdateMealRequest {
  food_name?: string
  meal_type?: number
  meal_date?: string
  image?: File
  calories?: number
  protein_g?: number
  fat_g?: number
  carbs_g?: number
}

export interface FoodStatistic {
  food_name: string
  count: number
}

export interface NutritionSummary {
  calories: number
  protein_g: number
  fat_g: number
  carbs_g: number
}

export interface DateRangeResult {
  records: Meal[]
  statistics: FoodStatistic[]
  nutrition_summary: NutritionSummary
}

export const mealService = {
  getMeals: async (): Promise<Meal[]> => {
    const response = await api.get<{ data: Meal[] }>('/api/v1/meals')
    return response.data.data
  },

  createMeal: async (data: CreateMealRequest): Promise<Meal> => {
    const formData = new FormData()
    formData.append('food_name', data.food_name)
    formData.append('meal_type', data.meal_type.toString())
    formData.append('meal_date', data.meal_date)
    if (data.image) formData.append('image', data.image)
    if (data.calories) formData.append('calories', data.calories.toString())
    if (data.protein_g) formData.append('protein_g', data.protein_g.toString())
    if (data.fat_g) formData.append('fat_g', data.fat_g.toString())
    if (data.carbs_g) formData.append('carbs_g', data.carbs_g.toString())

    const response = await api.post<{ data: Meal }>('/api/v1/meals', formData, {
      headers: { 'Content-Type': 'multipart/form-data' },
    })
    return response.data.data
  },

  updateMeal: async (id: number, data: UpdateMealRequest): Promise<Meal> => {
    const formData = new FormData()
    if (data.food_name) formData.append('food_name', data.food_name)
    if (data.meal_type) formData.append('meal_type', data.meal_type.toString())
    if (data.meal_date) formData.append('meal_date', data.meal_date)
    if (data.image) formData.append('image', data.image)
    if (data.calories !== undefined) formData.append('calories', data.calories.toString())
    if (data.protein_g !== undefined) formData.append('protein_g', data.protein_g.toString())
    if (data.fat_g !== undefined) formData.append('fat_g', data.fat_g.toString())
    if (data.carbs_g !== undefined) formData.append('carbs_g', data.carbs_g.toString())

    const response = await api.put<{ data: Meal }>(`/api/v1/meals/${id}`, formData, {
      headers: { 'Content-Type': 'multipart/form-data' },
    })
    return response.data.data
  },

  deleteMeal: async (id: number): Promise<void> => {
    await api.delete(`/api/v1/meals/${id}`)
  },

  getMealsByDateRange: async (startDate: string, endDate: string): Promise<DateRangeResult> => {
    const response = await api.get<{ data: DateRangeResult }>('/api/v1/meals/range', {
      params: { start_date: startDate, end_date: endDate },
    })
    return response.data.data
  },

  exportCSV: async (startDate: string, endDate: string): Promise<void> => {
    const token = localStorage.getItem('token')
    const baseUrl = import.meta.env.VITE_API_BASE_URL || 'http://localhost:8080'
    const response = await fetch(
      `${baseUrl}/api/v1/meals/export?start_date=${startDate}&end_date=${endDate}`,
      { headers: { Authorization: `Bearer ${token}` } }
    )
    if (!response.ok) throw new Error('导出失败')

    const blob = await response.blob()
    const url = URL.createObjectURL(blob)
    const a = document.createElement('a')
    a.href = url
    a.download = `meals_export_${startDate}_${endDate}.csv`
    document.body.appendChild(a)
    a.click()
    document.body.removeChild(a)
    URL.revokeObjectURL(url)
  },

  getImageUrl: (filename: string): string => {
    return `/images/${filename}`
  },
}
